File: 20242Q/poo/laboratorio/tp03/ej9/src/Triangle.java
public class Triangle extends Figure {

    Point p1, p2, p3;

    public Triangle(Point p1, Point p2, Point p3) {
        this.p1 = p1;
        this.p2 = p2;
        this.p3 = p3;
    }

    @Override
    public double area() {
        // fórmula de Herón
        double semiPerimeter = perimeter() / 2;
        return Math.sqrt(semiPerimeter * (semiPerimeter - firstSide()) * (semiPerimeter - secondSide()) * (semiPerimeter - thirdSide()));
    }

    @Override
    public double perimeter() {
        return firstSide() + secondSide() + thirdSide();
    }

    private double firstSide() {
        return Math.abs(p1.distanceTo(p2));
    }

    private double secondSide() {
        return Math.abs(p2.distanceTo(p3));
    }

    private double thirdSide() {
        return Math.abs(p3.distanceTo(p1));
    }

    @Override
    public String toString() {
        return "Triángulo [ %s , %s , %s ]".formatted(p1, p2, p3);
    }
}

File: 20242Q/poo/laboratorio/tp08/ej4/triangle.rb
require_relative 'figure'

class Triangle < Figure
  def initialize(first, second, third)
    @first = first
    @second = second
    @third = third
  end

  def perimeter
    first_side + second_side + third_side
  end

  def area
    semiperimeter = perimeter / 2
    Math.sqrt(semiperimeter * (semiperimeter - first_side) * (semiperimeter - second_side) * (semiperimeter - third_side))
  end

  def to_s
    "Triángulo [ #{@first} , #{@second} , #{@third} ]"
  end

  private

  def first_side
    @first.distance(@second)
  end

  def second_side
    @second.distance(@third)
  end

  def third_side
    @third.distance(@first)
  end
end
